### PR TITLE
Exclude test files in virtual env or workspace cache

### PR DESCRIPTION
**Pull request by RaymonGulati1**
Issue #5488

### CODE REVIEW

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 1]
```diff
@@ -173,7 +176,9 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
                 _testFilesUpdateWatcher.AddDirectoryWatch(workspace.Location);
                 oldWatcher?.Dispose();
 
-                var files = Directory.EnumerateFiles(workspace.Location, "*.*", SearchOption.AllDirectories);
+                var files = Directory.EnumerateFiles(workspace.Location, "*.py", SearchOption.AllDirectories)
```
Comment (huguesv) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: I think you should save the results of `interpreterRegistryService.Configurations.Where(x => PathUtils.IsSubpathOf(projectInfo.ProjectHome, x.InterpreterPath))` here in a variable and reuse it, so you don't have to recalculate that for potentially 100s or 1000s of files.

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 2]
```diff
@@ -173,7 +176,9 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
...
+                    .Where(x => !IsFileExcluded(projInfo, x)
+                );
```
Comment (bschnurr) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Did you consider using EnumerateDirectories and checking for InterpreterPath? 
Maybe a more common case that enivronments folder is one of the folders directory under workspace.
then calling EnumerateFiles on the remaining folders.. 
I would assume its a faster then checking every file.

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs (hunk 1 above)
Comment (RaymonGulati1) [reply Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Updated

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs (hunk 2 above)
Comment (RaymonGulati1) [reply Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Updated

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 3]
```diff
@@ -173,8 +176,7 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
                 _testFilesUpdateWatcher.AddDirectoryWatch(workspace.Location);
                 oldWatcher?.Dispose();
 
-                var files = Directory.EnumerateFiles(workspace.Location, "*.*", SearchOption.AllDirectories);
-                foreach (var file in files) {
+                foreach (var file in FilterInvalidFiles(projInfo)) {
```
Comment (huguesv) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: You should call it `GetContainerFiles`

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 4]
```diff
@@ -288,5 +290,31 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
             }
             NotifyContainerChanged();
         }
+
+        private IEnumerable<string> FilterInvalidFiles(ProjectInfo projectInfo) {
+            List<string> validFiles = Directory.EnumerateFiles(projectInfo.ProjectHome, "*.py", SearchOption.TopDirectoryOnly).ToList();
```
Comment (huguesv) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: use var for both

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 5]
```diff
@@ -288,5 +290,31 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
...
+            List<InterpreterConfiguration> workspaceInterpreterFactories = _interpreterRegistryService.Configurations
+                .Where(x => PathUtils.IsSubpathOf(projectInfo.ProjectHome, x.InterpreterPath))
+                .ToList();
+
+            foreach (var directory in Directory.EnumerateDirectories(projectInfo.ProjectHome, "*", SearchOption.TopDirectoryOnly)) {
```
Comment (huguesv) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Those are the defaults, you can just do `Directory.EnumerateDirectories(projectInfo.ProjectHome)`

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 6]
```diff
@@ -288,5 +290,31 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
...
+        private IEnumerable<string> GetTestFiles(ProjectInfo projectInfo) {
+            List<string> validFiles = Directory.EnumerateFiles(projectInfo.ProjectHome, "*.py", SearchOption.TopDirectoryOnly).ToList();
+            List<InterpreterConfiguration> workspaceInterpreterFactories = _interpreterRegistryService.Configurations
+                .Where(x => PathUtils.IsSubpathOf(projectInfo.ProjectHome, x.InterpreterPath))
+                .ToList();
+
+            foreach (var directory in Directory.EnumerateDirectories(projectInfo.ProjectHome, "*", SearchOption.TopDirectoryOnly)) {
+
+                if (!workspaceInterpreterFactories.Any(x => PathUtils.IsSameDirectory(x.GetPrefixPath(), directory)) &&
+                    !PathUtils.IsSameDirectory(directory, Path.Combine(projectInfo.ProjectHome, ".vs"))
```
Comment (bschnurr) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: could move this outside the loop. Path.Combine(projectInfo.ProjectHome, ".vs")

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 7]
```diff
@@ -288,5 +290,31 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
             }
             NotifyContainerChanged();
         }
+
+        private IEnumerable<string> GetTestFiles(ProjectInfo projectInfo) {
+            List<string> validFiles = Directory.EnumerateFiles(projectInfo.ProjectHome, "*.py", SearchOption.TopDirectoryOnly).ToList();
```
Comment (bschnurr) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Does validFiles.AddRange still work if we leave it as an IEnumerable?  ie. remove the ToList().

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 8]
```diff
@@ -288,5 +290,31 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
             }
             NotifyContainerChanged();
         }
+
+        private IEnumerable<string> FilterInvalidFiles(ProjectInfo projectInfo) {
+            List<string> validFiles = Directory.EnumerateFiles(projectInfo.ProjectHome, "*.py", SearchOption.TopDirectoryOnly).ToList();
+            List<InterpreterConfiguration> workspaceInterpreterFactories = _interpreterRegistryService.Configurations
```
Comment (huguesv) [inline Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: They're configurations, not factories, so I'd call it `workspaceConfigs`.

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs (hunk 7 above)
Comment (RaymonGulati1) [reply Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: There is no AddRange for IEnumerable, so I've replaced it with [IEnumerable.Concat()](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.concat?view=netframework-4.8).

**Pull request merged**